sync/server.md

```diff
@@ -0,0 +1,404 @@
+See [Sync UI](./ui.md) for more info on the Admin UI Architecture
+
+## PostgreSQL Connector
+
+### Data Source
+
+The defintion for a PostgreSQL Data Source will have the following parameters:
+
+* Database Server Host
+* Database Server Port
+* Database Name
+* Database Username
+* Database Password
+
+Table names will not be configured as part of the Data Source for Postgres.
+Instead, the table will be determined by the SQL query defined in the Resolver Request Mapping.
+
+### Resolver Request Mapping
+
+This is the defintion of an SQL query to execute.
+Any parameters defined in the Data Schema for the Query or Mutation being executed will be made available on the `conext` object as `context.arguments`.
+For example, to find a single record by id, use the following:
+
+```
+// readNote(id: String): Note
+
+SELECT id, title, content, timestamp FROM \"Notes\" WHERE id='{{context.arguments.id}}';
+```
+
+To update a row, use the following:
+
+```
+// updateNote(id: String,title: String): Note
+
+UPDATE \"Notes\" SET \"title\"='{{context.arguments.title}}' WHERE \"id\"='{{context.arguments.id}}' RETURNING *;
+```
+
+### Resolver Response Mapping
+
+This is the defintion for how to format the result of a SQL query to match the expected GraphQL Data Schema for that Query or Mutation.
+The result of the SQL query will be made available on the `context.result` object.
+The `result` field will always be an `array`.
+If a list of objects is expected in the Schema rsponse, this can be mapped as follows:
+
+```
+{{toJSON context.result}}
+```
+
+If a max of 1 result is expected, e.g. finding a record by id, this can be mapped as follows:
+
+```
+{{toJSON context.result.[0]}}
+```
+
+## HTTP Connector
+
+### Data Source
+
+The defintion of a HTTP Data Source will have the following parameters:
+
+* Protocol i.e. HTTP or HTTPS
+* Host e.g. www.example.com
+* Port
+* Base path e.g. /myapi/v1
+* Headers (optional)
+* Basic Auth Username (optional)
+* Basic Auth Password (optional)
+
+### Resolver Request Mapping
+
+This is the defintion of a HTTP request to make for a specific Query or Mutation.
+It will have the following parameters in conjunciton with the Data Source parameters to make a request.
+
+* Path (Added to the end of the Base path defined in the Data Source. Can include query parameters)
+* Additional Headers (optional)
+* Method e.g. GET, POST, PUT
+* Request Body (Optional)
+
+These will be defined as a JSON object, but will be treated as a string.
+This means the below defintion would actually be stored as the string beneath
+
+JSON version
+```
+{
+  "path": "/notes",
+  "method": "GET",
+  "headers": {
+    "accept": "application/json"
+  }
+}
+```
+
+String version
+```
+{\n  \"path\": \"/notes\",\n  \"method\": \"GET\",\n  \"headers\": {\n    \"accept\": \"application/json\"\n  }}
+```
+
+By treating the defintion as a string, it allows for Handlebars syntax to be added anywhere in the JSON object.
```

Comment:
@wei-lee you expressed some concern about the limitations of handlebars templates.
If we define exactly what we want developers to do in a request/response mapping, we should be able to choose the most appropriate template engine.
This site is handy for narrowing down what engine to use
http://garann.github.io/template-chooser/

My thinking is we want to limit what's possible in request/response mappings to data manipulation e.g. iterating, filtering, mapping.
For more complex logic, let the develop do that outside of data sync in their own service (and use the http data source), or hook up to OpenWhisk actions for custom code.
I have security concerns if we bring in a more flexible template engine that allows js code execution.

cc @johnfriz for thoughts

Reply:
I'm thinking the following:

* can use template placeholders for arguments/results i.e. can access `context.arguments` & `context.result`
* can iterate through an array of results, filtering out entries based on some conditional (for authorisation)
* can iterate through an array of results, mapping entries from one structure to another (this may be more useful when dealing with non json data e.g. an xml doc
* can stringify json